ocaml/xapi/nm.ml

```diff
@@ -322,54 +322,68 @@ let bring_pif_up ~__context ?(management_interface=false) (pif: API.ref_PIF) =
         Db.PIF.set_currently_attached ~__context ~self:pif ~value:currently_attached
       | _ ->
         let dbg = Context.string_of_task __context in
-        let net_rc = Db.Network.get_record ~__context ~self:rc.API.pIF_network in
+        let net_rc = Db.Network.get_record ~__context ~self:!rc.API.pIF_network in
         let bridge = net_rc.API.network_bridge in
 
         (* Call networkd even if currently_attached is false, just to update its state *)
-        debug "Making sure that PIF %s is up" rc.API.pIF_uuid;
+        debug "Making sure that PIF %s is up" !rc.API.pIF_uuid;
 
         (* If the PIF is a bond master, the bond slaves will now go down *)
         (* Interface-reconfigure in bridge mode requires us to set currently_attached to false here *)
-        begin match rc.API.pIF_bond_master_of with
+        begin match !rc.API.pIF_bond_master_of with
           | [] -> ()
           | bond :: _ ->
-            let slaves = Db.Bond.get_slaves ~__context ~self:bond in
+            let dbBond = Db.Bond.get_record ~__context ~self:bond in
+
+            if dbBond.API.bond_auto_update_mac then begin
+              (* Make sure the MAC address of the PIF Bond master matches the MAC address of the Bond's primary slave *)
+              let primarySlaveRef = dbBond.API.bond_primary_slave in
+              let primarySlaveMAC = Db.PIF.get_MAC ~__context ~self:primarySlaveRef in
+              if !rc.pIF_MAC <> primarySlaveMAC then begin
+                Db.PIF.set_MAC ~__context ~self:pif ~value:primarySlaveMAC;
+                rc := Db.PIF.get_record ~__context ~self:pif;
```

Comment:
the reason i made `rc` a ref was so that we can conditionally make this `get_record` call. would you say the cost of introducing ref is worth this small benefit?

Reply:
I think that adding a new control variable would be better than changing rc to be a reference, as the latter introduces quite a few avoidable code changes.

For example:
```
let mac_changed = ref false in

[...]

if !rc.pIF_MAC <> primarySlaveMAC then
  mac_changed := true;

[...]

let rc = if !mac_changed then [...] else rc in
```

This keeps the changes localised and it's very easy to understand what happens (change the name of the variable or add a comment if it's not clear)